### DISCUSSION TITLE
Mod tracker music playback

### DISCUSSION THREAD

**Discussion by nilskj** [excaliburjs/Excalibur#2058] [Ideas]
Great work with this Fantastic engine, great ease of use! Would be nice to also enable playing mod tracker files (.xm .s3m .mod). Goes well with retro style games and with tiny filesizes they are great for the web context.

I tried some other libs but most of them are very old and works poorly. This would be very convenient to have in the Sound class... or even better as an opt in separate library though and just write a ModSound class that wraps it up, extending Sound

**Comment by eonarheim** [comment]
Hi @nilskj thanks for reaching out and the kind words!

Totally, this sounds like a great feature to add as a plugin! I'm not familiar with mod tracker files but I'll add this to our radar ([roadmap](https://github.com/excaliburjs/Excalibur/issues/1161)).

It should be possible in theory to parse them and produce sound with the underlying WebAudio API. They seem similar in concept to MIDI (which I think would also be cool to support in some way)

Have you found any JS libraries that you like to parse/play MOD tracker files?

**Reply by eonarheim** [reply]
This library looks like it could be promising https://github.com/steffest/BassoonTracker

**Reply by nilskj** [reply]
Great! This is another one I tried and got it to work after some struggle https://github.com/electronoora/webaudio-mod-player

**Reply by nilskj** [reply]
Since I needed this quickly for a game jam I built a wrapper around webaudio-mod-player that can just be plugged and played with `npm install`. [microjazz](https://github.com/nilskj/microjazz). It is a working re-implementation so next step is to build a custom Sound derived class that can link it up. Is there a plugin system or how do you feel about this approach?

To use it with excaliburjs already now: Provide a file url that you would like to load. If using a bundler you can put the tunes in /static or /public (ex for Vite) folder to fetch them. 

```javascript
    import { ModPlayer } from "microjazz"

    const player = new ModPlayer()
    player.load("awesome-chiptune.mod")
    player.onReady = () => {
        player.play()
    }
```
This is all I had time for right now, it could be split up better for tree shaking and minified.
